src/interface/swarm.cpp

```diff
@@ -0,0 +1,345 @@
+  pmy_block->par_for(
+      "setPoolMax_mark_1", 0, nmax_pool_ - 1,
+      KOKKOS_LAMBDA(const int n) { newvar_bool_data(n) = oldvar_bool_data(n); });
+  pmy_block->par_for(
+      "setPoolMax_mark_2", nmax_pool_, nmax_pool - 1,
+      KOKKOS_LAMBDA(const int n) { newvar_bool_data(n) = 0; });
+  marked_for_removal_ = newvar_bool;
+
+  // TODO this is not an efficient loop ordering, probably
+  for (int n = 0; n < intVector_.size(); n++) {
+    auto oldvar = intVector_[n];
+    auto newvar = std::make_shared<ParticleVariable<int>>(oldvar->label(), nmax_pool,
+                                                          oldvar->metadata());
+    auto oldvar_data = oldvar->data;
+    auto newvar_data = newvar->data;
+    pmy_block->par_for(
+        "setPoolMax_int", 0, nmax_pool_ - 1,
+        KOKKOS_LAMBDA(const int m) { newvar_data(m) = oldvar_data(m); });
+
+    intVector_[n] = newvar;
+    intMap_[oldvar->label()] = newvar;
+  }
+
+  for (int n = 0; n < realVector_.size(); n++) {
+    auto oldvar = realVector_[n];
+    auto newvar = std::make_shared<ParticleVariable<Real>>(oldvar->label(), nmax_pool,
+                                                           oldvar->metadata());
+    auto oldvar_data = oldvar->data;
+    auto newvar_data = newvar->data;
+    pmy_block->par_for(
+        "setPoolMax_real", 0, nmax_pool_ - 1,
+        KOKKOS_LAMBDA(const int m) { newvar_data(m) = oldvar_data(m); });
+    realVector_[n] = newvar;
+    realMap_[oldvar->label()] = newvar;
+  }
+
+  nmax_pool_ = nmax_pool;
+}
+
+ParArrayND<bool> Swarm::AddEmptyParticles(int num_to_add) {
+  while (free_indices_.size() < num_to_add) {
+    increasePoolMax();
+  }
+
+  ParArrayND<bool> new_mask("Newly created particles", nmax_pool_);
+  auto new_mask_h = new_mask.GetHostMirror();
+  for (int n = 0; n < nmax_pool_; n++) {
+    new_mask_h(n) = false;
+  }
+
+  auto mask_h = mask_.data.GetHostMirror();
+  mask_h.DeepCopy(mask_.data);
+
+  auto free_index = free_indices_.begin();
+
+  // Don't bother sanitizing the memory
+  for (int n = 0; n < num_to_add; n++) {
+    mask_h(*free_index) = true;
+    new_mask_h(*free_index) = true;
+    max_active_index_ = std::max<int>(max_active_index_, *free_index);
+
+    free_index = free_indices_.erase(free_index);
+  }
+
+  num_active_ += num_to_add;
+
+  new_mask.DeepCopy(new_mask_h);
+  mask_.data.DeepCopy(mask_h);
+
+  return new_mask;
+}
+
+// No active particles: nmax_active_index = -1
+// No particles removed: nmax_active_index unchanged
+// Particles removed: nmax_active_index is new max active index
+void Swarm::RemoveMarkedParticles() {
+  int new_max_active_index = -1; // TODO BRR this is a magic number, needed for Defrag()
+
+  auto mask_h = mask_.data.GetHostMirrorAndCopy();
+  auto marked_for_removal_h = marked_for_removal_.data.GetHostMirror();
+  marked_for_removal_h.DeepCopy(marked_for_removal_.data);
+
+  // loop backwards to keep free_indices_ updated correctly
+  for (int n = max_active_index_; n >= 0; n--) {
```

Comment:
At some point, it might be worth thinking about how this can be done on the device.

Reply:
Yes I agree. There may be some game to play with the outer/inner par_for loop model used elsewhere? Some sort of recursive parallel algorithm that evaluates/sorts/merges little linked lists together into one big one.

Reply:
We could also investigate `Kokkos` native data structures that have parallel-safe properties for this kind of thing. Like atomic operations, maps, lists, etc

Reply:
Anyway, it's for the future. I wouldn't do it now.